src/manager/node.c

```diff
@@ -334,38 +380,18 @@ static int node_match_unit_new(sd_bus_message *m, void *userdata, UNUSED sd_bus_
                         free(usubs->substate);
                         usubs->substate = strdup("running");
                 }
-                UnitSubscription *usub = NULL;
-                UnitSubscription *next_usub = NULL;
-                LIST_FOREACH_SAFE(subs, usub, next_usub, usubs->subs) {
-                        Subscription *sub = usub->sub;
-                        int r = sub->handle_unit_new(sub->monitor, node->name, unit, reason);
-                        if (r < 0) {
-                                hirte_log_errorf("Failed to emit UnitNew signal: %s", strerror(-r));
-                        }
-                }
         }
 
-        /* Explicitly check for wildcard subscriptions, but only when the provided unit in the signal isn't
-         * one. That case is already handled by the regular logic above. */
-        if (!is_wildcard(unit)) {
-                const UnitSubscriptionsKey wildcard_key = { (char *) SYMBOL_WILDCARD };
-                UnitSubscriptions *usubs_wildcard = hashmap_get(node->unit_subscriptions, &wildcard_key);
-                if (usubs_wildcard != NULL) {
-                        UnitSubscription *usub = NULL;
-                        UnitSubscription *next_usub = NULL;
-                        LIST_FOREACH_SAFE(subs, usub, next_usub, usubs_wildcard->subs) {
-                                Subscription *sub = usub->sub;
-
-                                /* only emit signal once per monitor */
-                                if (unit_subscriptions_contains_monitor(usubs, (Monitor *) sub->monitor)) {
-                                        continue;
-                                }
-
-                                int r = sub->handle_unit_new(sub->monitor, node->name, unit, reason);
-                                if (r < 0) {
-                                        hirte_log_errorf("Failed to emit UnitNew signal: %s", strerror(-r));
-                                }
-                        }
+        Monitor *unique_monitor = NULL;
+        Monitor *unique_monitor_next = NULL;
+        LIST_FOREACH_SAFE(
+                        monitors,
+                        unique_monitor,
+                        unique_monitor_next,
+                        node_compute_unique_monitor_list(node, unit)) {
```

Comment:
Is this really right if `unit == "*"`? I'm not sure. Seems like we may sometimes send both virual new "*" from the manager and then the agent in some cases. Needs more careful analysis.

Reply:
I think this should be alright for `*`.  The agent sends only a virtual UnitNew for wildcards when such a subscription is made - which is forwarded by the manager to the monitor. Since wildcard units and "regular" units are stored in the `unit_subscriptions` of the respective node, `node_compute_unique_monitor_list` prevents duplicate emitting the UnitNew signal to the monitor. 
If the node disconnects, then the monitor with the wildcard gets treated the same as regular unit subs and a StateChanged (with the `agent-offline`) substate as well as a UnitRemoved gets emitted for it. On a reconnect of the agent, the manager subscribes with the wildcard again and we receive another UnitNew signal. 
Here is a sample output for subscribing, disconnecting the agent and reconnecting again:
```
Monitor path: /org/containers/hirte/monitor/5
Subscribing to node 'laptop' and unit '*'
[laptop] *
        Unit created (reason: virtual)
[laptop] *
        Unit state changed (reason: virtual)
        Active: inactive (agent-offline)
[laptop] *
        Unit removed (reason: virtual)
[laptop] *
        Unit created (reason: virtual)

```